### PR TITLE
Stripe products prices

### PR DESCRIPTION
**Pull request by dparker1005**
### All Submissions:

* [x] Have you followed the [Contributing guideline](https://github.com/strangerstudios/paid-memberships-pro/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/strangerstudios/paid-memberships-pro/pulls) for the same update/change?

### Changes proposed in this Pull Request:
Start using a single Stripe product per level and a single Stripe Price per payment amount.

Also contains changes to make the `process` method easier to follow for each page load and reduce "magic" properties stored in the MemberOrder.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

File: classes/gateways/class.pmprogateway_stripe.php [hunk 1]
```diff
@@ -1292,52 +1279,120 @@ public static function dependencies() {
...
+		$payment_transaction_id = '';
+		$subscription_transaction_id = '';
+
+		if ( ! empty( $order->setup_intent_id ) ){
+			// The user tried to confirm a setup intent. This means that there was no initial
+			// payment needed for this chekout (otherwise there would be a payment intent instead),
+			// and that the subscription needed authorization before being created.
+			$setup_intent = $this->process_setup_intent( $order->setup_intent_id );
+			if ( is_string( $setup_intent ) ) {
+				$order->error      = __( "Error processing setup intent.", 'paid-memberships-pro' ) . " " . $setup_intent;
```
Comment (sc0ttkclark) [inline classes/gateways/class.pmprogateway_stripe.php:1291]:
```suggestion
				$order->error      = __( 'Error processing setup intent.', 'paid-memberships-pro' ) . ' ' . $setup_intent;
```

File: classes/gateways/class.pmprogateway_stripe.php [hunk 2]
```diff
@@ -1292,52 +1279,120 @@ public static function dependencies() {
...
+		} else {
+			// User has either just submitted the checkout form or tried to confirm their
+			// payment intent.
+			$customer = null; // This will be used to create the subscription later.
+			if ( ! empty( $order->payment_intent_id ) ) {
+				// User has just tried to confirm their payment intent. We need to make sure that it was
+				// confirmed successfully, and then try to create their subscription if needed.
+				$payment_intent = $this->process_payment_intent( $order->payment_intent_id );
+				if ( is_string( $payment_intent ) ) {
+					$order->error      = __( "Error processing payment intent.", 'paid-memberships-pro' ) . " " . $payment_intent;
```
Comment (sc0ttkclark) [inline classes/gateways/class.pmprogateway_stripe.php:1306]:
```suggestion
					$order->error      = __( 'Error processing payment intent.', 'paid-memberships-pro' ) . ' ' . $payment_intent;
```

File: classes/gateways/class.pmprogateway_stripe.php [hunk 3]
```diff
@@ -1292,52 +1279,120 @@ public static function dependencies() {
...
-		if ( is_string( $customer ) ) {
-			// There was an issue updating the default payment method.
-			$order->error      = __( "Error updating default payment method.", 'paid-memberships-pro' ) . " " . $customer;
-			$order->shorterror = $order->error;
-			return false;
-		}
+				$payment_method = $this->get_payment_method( $order );
+				if ( empty( $payment_method ) ) {
+					// There was an issue getting the payment method.
+					$order->error      = __( "Error retrieving payment method.", 'paid-memberships-pro' );
```
Comment (sc0ttkclark) [inline classes/gateways/class.pmprogateway_stripe.php:1328]:
```suggestion
					$order->error      = __( 'Error retrieving payment method.', 'paid-memberships-pro' );
```

File: classes/gateways/class.pmprogateway_stripe.php [hunk 4]
```diff
@@ -1292,52 +1279,120 @@ public static function dependencies() {
...
+				}
 
-		// Save customer and payment method in $order.
-		// This will likely be removed as we rework payment processing.
-		$order->stripe_customer = $customer;
-		$order->stripe_payment_method = $payment_method;
+				$customer = $this->set_default_payment_method_for_customer( $customer, $payment_method );
+				if ( is_string( $customer ) ) {
+					// There was an issue updating the default payment method.
+					$order->error      = __( "Error updating default payment method.", 'paid-memberships-pro' ) . " " . $customer;
```
Comment (sc0ttkclark) [inline classes/gateways/class.pmprogateway_stripe.php:1336]:
```suggestion
					$order->error      = __( 'Error updating default payment method.', 'paid-memberships-pro' ) . ' ' . $customer;
```

File: classes/gateways/class.pmprogateway_stripe.php [hunk 5]
```diff
@@ -1292,52 +1279,120 @@ public static function dependencies() {
...
+					$payment_transaction_id = $order->stripe_payment_intent->charges->data[0]->id;
+				}
+			}
+
+			// Create a subscription if we need to.
+			if ( pmpro_isLevelRecurring( $order->membership_level ) ) {
+				$subscription = $this->create_subscription_for_customer_from_order( $customer->id, $order );
+				if ( empty( $subscription ) ) {
+					// There was an issue creating the subscription.
+					$order->error      = __( "Error creating subscription for customer.", 'paid-memberships-pro' );
```
Comment (sc0ttkclark) [inline classes/gateways/class.pmprogateway_stripe.php:1364]:
```suggestion
					$order->error      = __( 'Error creating subscription for customer.', 'paid-memberships-pro' );
```

**Review by sc0ttkclark — CHANGES_REQUESTED** [review]
Some quick changes here, looking good so far. I'll do more review later too

**Review by sc0ttkclark — APPROVED** [review]

File: classes/gateways/class.pmprogateway_stripe.php [hunk 6]
```diff
@@ -2032,30 +2082,420 @@ private function convert_price_to_unit_amount( $price ) {
...
+			}
+		}
+
+		if ( empty( $level->ID ) ) {
+			// We do not have a valid level.
+			return;
+		}
+
+		$product_args = array(
+			'name' => $level->name,
```
[the remark this answers is not in the document]
Comment (ideadude) [reply classes/gateways/class.pmprogateway_stripe.php:2261]: The reference doesn't mention anything. https://stripe.com/docs/api/products/create#create_product-name

Playing around with names and description in the Stripe dashboard, it seems like any unicode character is allowed, including emojis, etc. When I tried to add HTML, it was included but not rendered in the Stripe dashboard. So they might be escaping that.

I think for now, let's trust Stripe to handle it.

**Pull request merged**